src/lib/err/backtrace.c

```diff
@@ -235,6 +238,24 @@ install_bt_handler(const char *software)
 static void
 remove_bt_handler(void)
 {
+  int i;
+
+  struct sigaction sa;
+
+  memset(&sa, 0, sizeof(sa));
+  sa.sa_handler = SIG_DFL;
+  sigfillset(&sa.sa_mask);
+
+  for (i = 0; trap_signals[i] >= 0; ++i) {
+    /* remove_bt_handler() is called on shutdown, from low-level code.
+     * It's not a fatal error, so we just ignore it. */
+    (void)sigaction(trap_signals[i], &sa, NULL);
+  }
+
+  /* We can't use tor_mutex_free() here, because lock uses err.
+   * This function is called while err is being shut down, so we can't
+   * log or take any useful action if pthread_mutex_destroy() fails. */
+  (void)pthread_mutex_destroy(&cb_buf_mutex);
```

Comment:
Is this comment accurate? It seems to me it is saying we can be in this handler while the mutex is locked? I would be a bit surprised but just in case, this can't be done. Man page states:

> It shall be safe to destroy an initialized mutex that is unlocked.  Attempting to destroy a locked mutex  or  a  mutex that  is  referenced  (for  example, while being used in a pthread_cond_timedwait() or pthread_cond_wait()) by another thread results in undefined behavior.

Reply:
You're right, there are two race conditions here:
1. log_backtrace_impl() can be called from other threads, and lock cb_buf_mutex, while the main thread is calling this function during shut down, which leads to undefined behaviour

2. cb_buf is used in a lot of functions, which could be called concurrently:
a) install_bt_handler
b) dump_stack_symbols_to_error_fds
c) crash_handler
d) log_backtrace_impl (protected by cb_buf_mutex)
Which could lead to corrupted stack traces.

I don't know if there is an easy way to fix this. Maybe I should just open a ticket.

@nmathewson do you have any ideas?